core/authors/views.py

```diff
@@ -184,4 +189,129 @@ def check_is_following(self, request, pk):
+            return Response("Size must be positive", status=400)
+
+        # post_types will track what level of posts a user can see
+        post_types = ["PUBLIC", "SERVERONLY"]
+        # convert to dict for dat O(1)
+        # Note: this is terrible, we should be using the database more directly
+        requesterFriends = {}
+        for friend in get_friends(requestingAuthor):
+            requesterFriends[friend] = True
+
+        # Check if they are direct friends
+        if requesterFriends.get(pk, False):
+            post_types += ["FRIENDS", "FOAF"]
+        else: # They are not direct friends, so we should check if they share any friends
+            for friend in get_friends(pk):
+                if requesterFriends.get(friend, False):
+                    post_types += ["FOAF"]
+                    break # we don't need to check any more friends
+
+        try:
+            posts = Posts.objects.all().filter(author=pk, visibility__in=post_types, unlisted=False)
+            # TODO: requestingAuthor is the one it should be visibleTo
+            #posts |= Posts.objects.all().filter(author=pk, visibility="PRIVATE", visibleTo=?????, unlisted=False)
+            posts.order_by('-published')
+        except:
+            print("got except!")
+            return Response(status=500)
+
+        pages = Paginator(posts, size)
+        posts = PostsSerializer(pages.page(page), many=True)
+
+        response = {
+            "query": "posts",
+            "count": pages.count,
+            "size": size,
+            # Recall: the page the user specifies is offset by +1 for Paginator
+            "next": "/author/{}/posts?page={}".format(pk,page) if page < pages.num_pages else None,
+            "previous": "/author/{}/posts?page={}".format(pk,page-2) if page > 1 else None,
+            "posts": posts.data
+        }
+        return Response(response, status=200)
+
+    # All posts visible to the currently auth'd user
+    # /author/posts
+    @action(detail=False, url_path="posts")
+    def posts(self, request):
+        if request.user.is_anonymous:
+            # TODO: should it be a text response?
+            return Response("You must be logged in to perform this action.",status=403)
+        
+        requestingAuthor = request.user.author.id # Should be guaranteed because we checked above
+        page = int(request.query_params.get("page", 0)) + 1 # Must offset page by 1
+        if page < 1:
+            return Response("Page number must be positive", status=400)
+
+        # TODO: size should be limited?
+        size = int(request.query_params.get("size", 50))
+        if size < 0:
+            return Response("Size must be positive", status=400)
+        
+        # Get direct friends and FOAFs into a dictionary
+        requesterFriends = {}
+        requesterFOAFs = {}
+        for friend in get_friends(requestingAuthor):
+            requesterFriends[friend] = True
+        for friend in requesterFriends:
```

Comment:
If you wanted to avoid making n queries you could do a join on the Follow object

SELECT *
FROM Follow a, Follow b
WHERE a.followed = b.follower
AND a.follower = b.followed

Reply:
That would return a list of all Friend relationships
Then you would either SELECT from that in an outer query, or iterate through the list through Python

Reply:
Sorry, this will get me all friends C that are FOAF of A?
I think that would require a triple join at least?

Reply:
The query I described would list all people who are friends with each other.
Then you could filter it out by doing:
for friend in results:
     if (requestFriends[friend.follower]):
         requesterFOAF[friend.followed] = True